yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryService.scala

```diff
@@ -0,0 +1,1048 @@
+      eventHandlingThread = Some(new Thread(new Dequeue(), "HistoryEventHandlingThread"))
+      eventHandlingThread.get.start()
+    } else {
+      logInfo("Timeline service is disabled")
+    }
+    if (registerListener()) {
+      logInfo(s"History Service listening for events: $this")
+    } else {
+      logInfo(s"History Service is not listening for events: $this")
+    }
+  }
+
+  /**
+   * Check the service configuration to see if the timeline service is enabled
+   * @return true if `YarnConfiguration.TIMELINE_SERVICE_ENABLED` is set.
+   */
+  def timelineServiceEnabled: Boolean = {
+    YarnTimelineUtils.timelineServiceEnabled(config)
+  }
+
+  /**
+   * Return a summary of the service state to help diagnose problems
+   * during test runs, possibly even production
+   * @return a summary of the current service state
+   */
+  override def toString(): String = {
+    s"YarnHistoryService for application $applicationId attempt $attemptId;" +
+    s" state=$serviceState;" +
+    s" endpoint=${_timelineWebappAddress};" +
+    s" bonded to ATS=$bondedToATS;" +
+    s" listening=$listening;" +
+    s" batchSize=$batchSize;" +
+    s" flush count=$getFlushCount;" +
+    s" total number queued=$eventsQueued, processed=$eventsProcessed;" +
+    s" attempted entity posts=$postAttempts" +
+    s" successful entity posts=$postSuccesses" +
+    s" failed entity posts=$postFailures;" +
+    s" events dropped=$eventsDropped;" +
+    s" app start event received=$appStartEventProcessed;" +
+    s" app end event received=$appEndEventProcessed;"
+  }
+
+  /**
+   * Is the service listening to events from the spark context?
+   * @return true if it has registered as a listener
+   */
+  def listening: Boolean = {
+    listener.isDefined
+  }
+
+  /**
+   * Is the service hooked up to an ATS server. This does not
+   * check the validity of the link, only whether or not the service
+   * has been set up to talk to ATS.
+   * @return true if the service has a timeline client
+   */
+  def bondedToATS: Boolean = {
+    _timelineClient.isDefined
+  }
+
+  /**
+   * Set the YARN binding information. This is called during startup. It is private
+   * to the package so that tests may update this data
+   * @param appId YARN application ID
+   * @param attemptId optional attempt ID
+   */
+  private[yarn] def bindToYarnApplication(appId: ApplicationId,
+      attemptId: Option[ApplicationAttemptId]): Unit = {
+    require(appId != null, "Null appId parameter")
+    _applicationId = appId
+    _attemptId = attemptId
+  }
+
+  /**
+   * Set the "spark" application and attempt information -the information
+   * provided in the start event. The attempt ID here may be `None`; even
+   * if set it may only be unique amongst the attempts of this application.
+   * That is: not unique enough to be used as the entity ID
+   * @param appId application ID
+   * @param attemptId attempt ID
+   */
+  private def setContextAppAndAttemptInfo(appId: Option[String],
```

Comment:
spark app ID comes in as "1", "2", "3", not a unique ID. Also, you get events coming in before the spark app start has  happened `SparkListenerEnvironmentUpdate`, to name one. It's not a bug, the one coming in in the context is ideal for UIs and subdirectories, and is used in some tests as its value is predictable. The YARN one has the strength of effectively being a UUID.

Currently the history publisher only uses that string value; it could take them as properties (spark.yarn.appId, spark.yarn.attemptId)?, if the back ends set them up right. Passing them down as the real YARN values provides the option of allowing other services to do more with them (i.e. if some other YARN/Hadoop APIs needed them, they'd be there) without having to rebuild them from strings, with all its brittleness. That said, putting them as properties would make it possible to pull the extension service code into spark core, and potentially allow the extension services to be deployed in the Mesos Scheduler.